multi/stm32l4-multi/tty.c

```diff
@@ -99,60 +142,209 @@ static int tty_irqHandler(unsigned int n, void *arg)
 {
 	tty_ctx_t *ctx = (tty_ctx_t *)arg;
 
-	if (*(ctx->base + isr) & ((1 << 5) | (1 << 3))) {
+	if ((*(ctx->base + isr) & ((1 << 5) | (1 << 3))) != 0) {
 		/* Clear overrun error bit */
 		*(ctx->base + icr) |= (1 << 3);
 
-		lf_fifo_push(&ctx->rx_fifo, *(ctx->base + rdr));
-		ctx->rxready = 1;
+		lf_fifo_push(&ctx->irq.rx_fifo, *(ctx->base + rdr));
+		ctx->irq.rxready = 1;
 	}
 
-	if (tty_txready(ctx))
+	if (tty_txready(ctx) != 0) {
 		*(ctx->base + cr1) &= ~(1 << 7);
+	}
+
+	return 1;
+}
+
+
+/* Nonnegative type must be one of dma_ht or dm_tc,
+   negative type means idle line detected in which case it's negated number of bytes read. */
+static void tty_dmaCallback(void *arg, int type)
+{
+	tty_ctx_t *ctx = (tty_ctx_t *)arg;
+	int rxready = TTY_RX_READY_FLAG;
+
+	if (type == dma_tc) {
+		rxready |= TTY_RX_FILLED_FLAG;
+		__atomic_fetch_add(&ctx->dma.debug.fill_cnt, 1, __ATOMIC_ACQ_REL);
```

Comment:
__atomic_* probably from C11.
Is that allowed?
Which standard is applicable C89, C99, C11?

Reply:
Yeah, it is from C11. We adhere to C99, but `__atomic_*` is really needed here. I see 3 solutions:
- use `__atomic_*`,
- use builtins
- write atomics from scratch.

Imho `__atomic_` is the lesser evil from all of them

Reply:
@badochov after some consideration I think it would be best to use C11 and `stdatomic.h` here